### ISSUE TITLE
memory leaky on DataLoader

### ISSUE THREAD

**Issue by leftthomas** [pytorch/pytorch#5812]
- OS: Ubuntu16.04
- PyTorch version: 0.3.1
- PyTorchNet version: 0.0.1
- How you installed PyTorch (conda, pip, source): conda
- Python version: 3.6.4
- CUDA/cuDNN version: CUDA 9.0.176/cuDNN 7.0.5.15

**The script to reproduce the bug:** 
``` python
from torch import nn
from torch.autograd import Variable
from torch.optim import Adam
from torchnet.engine import Engine
from tqdm import tqdm
import torchvision.transforms as transforms
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST

class MNISTNet(nn.Module):
    def __init__(self):
        super(MNISTNet, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=64, kernel_size=7, stride=2, padding=3),
            nn.Conv2d(in_channels=64, out_channels=128, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=128, out_channels=256, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=256, out_channels=512, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=512, out_channels=512, kernel_size=3, stride=1, padding=1)
        )
        self.classifier = nn.Linear(in_features=2 * 2 * 512, out_features=10)

    def forward(self, x):
        out = self.features(x)
        out = out.view(out.size(0), -1)
        classes = self.classifier(out)
        return classes


def get_iterator(mode):
    data = MNIST(root='./', train=mode, transform=transforms.ToTensor(), download=True)
    return DataLoader(dataset=data, batch_size=64, shuffle=mode, num_workers=4)


def processor(sample):
    data, labels, training = sample
    data = Variable(data).cuda()
    labels = Variable(labels).cuda()
    classes = model(data)
    loss = loss_criterion(classes, labels)
    return loss, classes


def on_sample(state):
    state['sample'].append(state['train'])


def on_start_epoch(state):
    state['iterator'] = tqdm(state['iterator'])


def on_end_epoch(state):
    engine.test(processor, get_iterator(False))


if __name__ == '__main__':

    model = MNISTNet().cuda()
    loss_criterion = nn.CrossEntropyLoss().cuda()

    optimizer = Adam(model.parameters())
    engine = Engine()
    engine.hooks['on_sample'] = on_sample
    engine.hooks['on_start_epoch'] = on_start_epoch
    engine.hooks['on_end_epoch'] = on_end_epoch

    engine.train(processor, get_iterator(True), maxepoch=100, optimizer=optimizer)
```
**Error messages**

- The `batch_size` of `DataLoader` is set to 64, then on epoch 1, before `on_end_epoch`:
![screenshot from 2018-03-15 20-23-31](https://user-images.githubusercontent.com/9991443/37463453-30928920-2890-11e8-93d0-d0dbad043997.png)

- After `on_end_epoch`, you could see the memory is grow up to about 3 times:
![screenshot from 2018-03-15 20-24-00](https://user-images.githubusercontent.com/9991443/37463507-5f8d2992-2890-11e8-9f7d-b6a754314c6a.png)

- Furthermore, if I change the `batch_size` of `DataLoader` to 1000, then on epoch 1, before `on_end_epoch`, you could see the memory use is just only **715MB**, it's strange:
<img width="867" alt="qq20180315-225225 2x" src="https://user-images.githubusercontent.com/9991443/37471232-3815a042-28a4-11e8-89e5-b971a9e3af93.png">

- After `on_end_epoch`, you could see the memory haven't change:
<img width="748" alt="qq20180315-225238 2x" src="https://user-images.githubusercontent.com/9991443/37471400-8e578920-28a4-11e8-8c51-34bb35c80e6c.png">

**Comment by fmassa** [comment]
I don' think this is a bug in the dataloader, because it doesn't send the data to the GPU.
It's probably a bug either in your code or in torchnet.

**Comment by leftthomas** [comment]
@fmassa but you could see from `processor` function that I have called `cuda()`, and according to  the docs of PyTorch
> Variable API is nearly the same as regular Tensor API (with the exception of a couple in-place methods, that would overwrite inputs required for gradient computation). In most cases Tensors can be safely replaced with Variables and the code will remain to work just fine. Because of this, we’re not documenting all the operations on variables, and you should refer to `torch.Tensor` docs for this purpose.

so it's fine by use `data = Variable(data).cuda()`, or it works only by call `data=Variable(data.cuda())`?

**Comment by leftthomas** [comment]
@fmassa I also remove `tnt` to test, and the problem still there,**The script to reproduce the bug:**
```python
from torch import nn
from torch.autograd import Variable
from torch.optim import Adam
from tqdm import tqdm
import torchvision.transforms as transforms
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST
import torch.nn.functional as F


class MNISTNet(nn.Module):
    def __init__(self):
        super(MNISTNet, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=64, kernel_size=7, stride=2, padding=3),
            nn.Conv2d(in_channels=64, out_channels=128, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=128, out_channels=256, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=256, out_channels=512, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=512, out_channels=512, kernel_size=3, stride=1, padding=1)
        )
        self.classifier = nn.Linear(in_features=2 * 2 * 512, out_features=10)

    def forward(self, x):
        out = self.features(x)
        out = out.view(out.size(0), -1)
        classes = self.classifier(out)
        return classes


if __name__ == '__main__':
    train_data = MNIST(root='../data/mnist', train=True, transform=transforms.ToTensor(), download=True)
    train_dl = DataLoader(dataset=train_data, batch_size=64, shuffle=True, num_workers=4)
    test_data =  MNIST(root='../data/mnist', train=False, transform=transforms.ToTensor(), download=True)
    test_dl = DataLoader(dataset=test_data, batch_size=64, num_workers=4)

    model = MNISTNet().cuda()
    optimizer = Adam(model.parameters())

    for epoch in tqdm(range(2)):
        for data,labels in train_dl:
            data = Variable(data).cuda()
            labels = Variable(labels).cuda()
            classes = model(data)
            loss = F.cross_entropy(classes, labels)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

        for data, labels in test_dl:
            data = Variable(data).cuda()
            labels = Variable(labels).cuda()
            classes = model(data)
```

**Comment by fmassa** [comment]
The reason in your last snippet is that the loss is still in scope when you do the testing loop, so it holds up the whole computation graph. You can either delete it before the testing starts `del loss` and `del classes`, or put the train loop in a separate function (and thus the scope ends with the function)

**Comment by leftthomas** [comment]
@fmassa I have tested **put the train loop in a separate function** as you said, it still have the problem! **It is a bug of PyTorch, not my code wrong**. The script to reproduce the bug:
```python
from torch import nn
from torch.autograd import Variable
from torch.optim import Adam
from tqdm import tqdm
import torchvision.transforms as transforms
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST
import torch.nn.functional as F


class MNISTNet(nn.Module):
    def __init__(self):
        super(MNISTNet, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=64, kernel_size=7, stride=2, padding=3),
            nn.Conv2d(in_channels=64, out_channels=128, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=128, out_channels=256, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=256, out_channels=512, kernel_size=3, stride=2, padding=1),
            nn.Conv2d(in_channels=512, out_channels=512, kernel_size=3, stride=1, padding=1)
        )
        self.classifier = nn.Linear(in_features=2 * 2 * 512, out_features=10)

    def forward(self, x):
        out = self.features(x)
        out = out.view(out.size(0), -1)
        classes = self.classifier(out)
        return classes


def train(model, optimizer):
    train_data = MNIST(root='./', train=True, transform=transforms.ToTensor(), download=True)
    train_dl = DataLoader(dataset=train_data, batch_size=64, shuffle=True, num_workers=4)

    for data, labels in train_dl:
        data = Variable(data).cuda()
        labels = Variable(labels).cuda()
        classes = model(data)
        loss = F.cross_entropy(classes, labels)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()


def test(model):
    test_data = MNIST(root='./', train=False, transform=transforms.ToTensor(), download=True)
    test_dl = DataLoader(dataset=test_data, batch_size=64, num_workers=4)
    for data, labels in test_dl:
        data = Variable(data).cuda()
        model(data)


if __name__ == '__main__':
    model = MNISTNet().cuda()
    optimizer = Adam(model.parameters())
    for epoch in tqdm(range(2)):
        train(model, optimizer)
        test(model)

```

**Comment by fmassa** [comment]
I tried your snippet and I don't see a memory leak, but a small increase in memory usage when the testing part happens, but it then stabilizes after the first epoch.

What I think is happening (but I might be wrong here) is that the caching allocator is not being able to recover part of the memory that was allocated during training, maybe because it got fragmented at a later stage of the network during training (for example during backprop) ?

In the end, I think the behavior you are seing is normal, and you'd see expected results if you cleared the caching allocator after training, but that would imply slowdowns on every epoch.

I think this is not a bug actually.

**Comment by leftthomas** [comment]
@fmassa But the memory grow up to three times! Does it really normal? And I have asked more than five friends to test this code, we have tested it on PyTorch 0.2.0, 0.3.0, 0.3.1, CUDA8.0,CUDA9.0, NVIDIA GTX 1060, 1070, 1080, TITAN X, they all have saw this issue. The most strange thing is when the `batch_size` set to other number such as 100, it won't encounter this problem, but if `batch_size`  is 32,64,128,256, the problem happened, I don't think it's fine, it's really the natural behavior of PyTorch?

**Comment by fmassa** [comment]
The memory bump that I saw in your example went from something like 780MB to 800MB, so not really significative I'd say.
About the difference in behavior on batch sizes, this might be cudnn choosing a different backend for some operations that requires different amount of memory.

**Comment by apaszke** [comment]
Also the value you're seeing in `nvidia-smi` is not the amount of memory *used* by PyTorch tensors at any given moment. PyTorch almost never frees the GPU memory because it's expensive and you can often reuse older allocations. As @fmassa you can use `empty_cache()` to get rid of the excessive use after every epoch if this is important for some reason.

**Comment by leftthomas** [comment]
@fmassa  @apaszke I saw a similar problem on [PyTorch discuss](https://discuss.pytorch.org/t/how-to-debug-causes-of-gpu-memory-leaks/6741) today.
<img width="798" alt="qq20180316-224408 2x" src="https://user-images.githubusercontent.com/9991443/37526893-9435bd04-296b-11e8-8630-c9d126275d6f.png">
As I said before, if the `batch_size` is not 64, the memory won't grow up, I mean **won't grow up**, not a little, and for 64, it bumps from **581MB** to **1739MB**, it's a huge memory usage.

**Comment by fmassa** [comment]
I'm sorry, I'm not sure what's going on on your system. I tried your snippet again without modification on PyTorch 0.3.1, and I only saw ~15BM more memory being allocated.
Are you running your latest script, the one that does not use `tnt`?

**Comment by leftthomas** [comment]
I tested the latest script I posted again now, and here are the screenshot, it still have this problem.
<img width="811" alt="qq20180316-232943 2x" src="https://user-images.githubusercontent.com/9991443/37530154-ea7d5caa-2973-11e8-89b4-cbea91305124.png">
<img width="874" alt="qq20180316-232953 2x" src="https://user-images.githubusercontent.com/9991443/37530184-f545145c-2973-11e8-8366-5f86af86eba4.png">
You could see from that the memory bumps from **581MB** to **1739MB**.

**Comment by soumith** [comment]
@leftthomas I've tried reproducing this on multiple machines with v0.3.1 from conda, cannot reproduce it. Maybe it's a problem on your nvidia driver or something. Can you help reproduce this in nvidia-docker so that we can help you.

**Comment by leftthomas** [comment]
@soumith , I think maybe it's about network, I have encounter another problem when I use PyTorch `vision` to load `FashionMNIST` data, and it downloads `MNIST`, not `FashionMNIST`, I have proposed this problem [here](https://github.com/pytorch/vision/issues/445), and it been confirmed as the 302 redirection by our university's cache server, solved on [here](https://github.com/zalandoresearch/fashion-mnist/issues/107). So I think maybe when I install my computer's system, CUDA, and PyTorch(or other computers in our university, they all reproduce this memory leaky problem), between this period, some files have to be downloaded is wrong, because of the 302 redirection by our university's cache server?

**Comment by soumith** [comment]
@leftthomas that might explain things. what is output of `print(torch.__version__)`

**Comment by leftthomas** [comment]
@soumith , it's `0.3.1.post2`. But I also tested PyTorch 0.2.0, 0.3.0, 0.3.1 and even build from source on the master 0.4 version.

**Comment by philippHRO** [comment]
@leftthomas Just saying this, because it hasn't been brought up yet. I had a similiar problem like you. I set my num_workers to 2 and everything worked fine, albeit a little slower of course. Num_workers = 4 just used too much RAM for me.

**Comment by philippHRO** [comment]
@leftthomas Just saying this, because it hasn't been brought up yet. I had a similiar problem like you. I set my num_workers to 1 and it worked fine, albeit a little slower of course. But I didn't test for many epochs yet. maybe it's still accumulating memory, only slower.

**Comment by leftthomas** [comment]
@soumith It not network problem, I have tested it again, and I found the memory leaky problem happened on GTX 1080, but not on GTX TITAN X.
